diff -urN a/log/logread.c b/log/logread.c
--- a/log/logread.c	2016-08-21 16:58:39.279766331 +0800
+++ b/log/logread.c	2016-08-21 16:59:06.507690282 +0800
@@ -176,7 +176,7 @@
 		uloop_timeout_set(&retry, 1000);
 	} else {
 		uloop_fd_add(&sender, ULOOP_READ);
-		syslog(0, "Logread connected to %s:%s\n", log_ip, log_port);
+		syslog(LOG_WARNING, "Logread connected to %s:%s\n", log_ip, log_port);
 	}
 }
 
@@ -345,7 +345,7 @@
 			err = send(sender.fd, buf, strlen(buf), 0);
 
 		if (err < 0) {
-			syslog(0, "failed to send log data to %s:%s via %s\n",
+			syslog(LOG_WARNING, "failed to send log data to %s:%s via %s\n",
 				log_ip, log_port, (log_udp) ? ("udp") : ("tcp"));
 			uloop_fd_delete(&sender);
 			close(sender.fd);
