diff -urN a/channel.c b/channel.c
--- a/channel.c	2016-05-04 00:33:17.104716372 -0700
+++ b/channel.c	2016-05-04 00:19:42.492836089 -0700
@@ -193,8 +193,10 @@
 	struct iw_scan_req scanopt;
 	struct iwreq wrq;
 	static u_int8_t i = 0;
+	int ret = 0;
 
 	while (true) {
+rescan:
 		if (!pause_scanning) {
 			bzero(&scanopt, sizeof(scanopt));
 			scanopt.max_channel_time = conf.duration_on_foreign * 3 / 2;
@@ -212,7 +214,13 @@
 			wrq.u.data.flags |= IW_SCAN_THIS_FREQ;
 			strncpy(wrq.ifr_name, conf.ifname_2g, IFNAMSIZ);
 
-			ioctl(mon_2g, SIOCSIWSCAN, &wrq);
+			ret = ioctl(mon_2g, SIOCSIWSCAN, &wrq);
+			if (ret < 0) {
+				sleep(1);
+				i = (i + conf.num_channels_2g - 1) % conf.num_channels_2g;
+				goto rescan;
+			}
+				
 		}
 		if (whole_channels_scanning) 
 			sleep(conf.scanning_intval * conf.num_channels_2g);
@@ -227,8 +235,10 @@
 	struct iw_scan_req scanopt;
 	struct iwreq wrq;
 	static u_int8_t i = 0;
+	int ret = 0;
 
 	while (true) {
+rescan:
 		if (!pause_scanning) {
 			bzero(&scanopt, sizeof(scanopt));
 			scanopt.max_channel_time = conf.duration_on_foreign * 3 / 2;
@@ -248,7 +258,12 @@
 			wrq.u.data.flags |= IW_SCAN_THIS_FREQ;
 			strncpy(wrq.ifr_name, conf.ifname_5g, IFNAMSIZ);
 				
-			ioctl(mon_5g, SIOCSIWSCAN, &wrq);
+			ret = ioctl(mon_5g, SIOCSIWSCAN, &wrq);
+			if (ret < 0) {
+				sleep(1);
+				i = (i + conf.num_channels_5g - 2) % conf.num_channels_5g;
+				goto rescan;
+			}
 		}
 		if (whole_channels_scanning) 
 			sleep(conf.scanning_intval * conf.num_channels_5g / 2);
