diff -urN a/capture.c b/capture.c
--- a/capture.c	2016-05-26 00:46:35.287939565 -0700
+++ b/capture.c	2016-05-26 00:47:29.865535167 -0700
@@ -28,6 +28,7 @@
 #include <net/if.h>
 #include <arpa/inet.h>
 #include <err.h>
+#include <errno.h>
 
 #include "capture.h"
 #include "util.h"
@@ -45,8 +46,10 @@
 
 
 	if ((ioctl(fd, SIOCGIFINDEX, &req)) < 0) {
-		goto retry;
-		//err(1, "Interface %s not found", devname);
+		if (ENODEV == errno)
+			goto retry;
+		else
+			err(1, "Interface %s not found", devname);
 	}
 
 	if (req.ifr_ifindex < 0)
