diff -urN a/control.c b/control.c
--- a/control.c	2016-06-16 00:51:08.042677679 -0700
+++ b/control.c	2016-06-16 00:53:55.690215821 -0700
@@ -51,10 +51,14 @@
 	int len = strlen(cmd);
 	char new[len+1];
 	char* pos;
+	char retry_count = 3;
 
 	while (access(conf.control_pipe, F_OK) < 0) {
 		printf("Waiting for control pipe...\n");
 		sleep(1);
+		retry_count--;
+		if (!retry_count)
+			exit(0);
 	}
 
 	ctlpipe = open(conf.control_pipe, O_WRONLY);
